Source/Particles/Algorithms/KineticEnergy.H

```diff
@@ -29,21 +30,22 @@ namespace Algorithms{
      *
      * @return the kinetic energy of the particle (in S.I. units)
      */
+    template <typename T>
     AMREX_GPU_HOST_DEVICE AMREX_INLINE
-    amrex::ParticleReal KineticEnergy(
-        const amrex::ParticleReal ux, const amrex::ParticleReal uy, const amrex::ParticleReal uz,
-        const amrex::ParticleReal mass)
+    T KineticEnergy(
+        const T ux, const T uy, const T uz, const T mass)
```

Comment:
The particle velocity components would be of type `ParticleReal`. Won't this cause a problem since it now expects all inputs to be of the templated type `T` (which is specified to be `double`)?
Could this be:
```suggestion
        const amrex::ParticleReal ux, const amrex::ParticleReal uy, const amrex::ParticleReal uz, const T mass)
```

Reply:
It should be fine if you force `double` precision, since it would be either a no-op or a promotion. But you are right, it could be an issue if for some reason `amrex::ParticleReal` is `double` and you try to force single precision.

Reply:
This could be a solution that avoids this issue:

```
    /**
     * \brief Computes the kinetic energy of a particle.
     * This method should not be used with photons.
     *
     * @tparam T floating-point type to be used in internal calculations (by default equalt to amrex::ParticleReal)
     * @param[in] ux x component of the particle momentum (code units)
     * @param[in] uy y component of the particle momentum (code units)
     * @param[in] uz z component of the particle momentum (code units)
     * @param[in] mass mass of the particle (in S.I. units)
     *
     * @return the kinetic energy of the particle (in S.I. units)
     */
    template <typename T = amrex::ParticleReal>
    AMREX_GPU_HOST_DEVICE AMREX_INLINE
    T KineticEnergy(
        const amrex::ParticleReal ux, const amrex::ParticleReal uy, const amrex::ParticleReal uz, const amrex::ParticleReal mass)
    {
        using namespace amrex;

        constexpr auto one = static_cast<T>(1.0);
        constexpr auto inv_c2 = one/Math::powi<2>(static_cast<T>(PhysConst::c));

        // The expression used is derived by reducing the expression
        // (gamma - 1)*(gamma + 1)/(gamma + 1)

        const auto u2 = static_cast<T>(ux*ux + uy*uy + uz*uz);
        const auto gamma = std::sqrt(one + u2*inv_c2);
        return one/(one + gamma)*mass*u2;
    }
```
`T` is by default equal to `ParticleReal`. However, you can override this and then calculations are carried out with precision `T` and the result is of type `T`. Conversions are carried out using `static_cast`, so it should work ok.